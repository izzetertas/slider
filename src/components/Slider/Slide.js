import React from 'react'
import './Slider.css'

const Slide = ({ content, index }) => (
  <div className='slider-item' key={index}>
    {content}
  </div>
)

export default Slide
