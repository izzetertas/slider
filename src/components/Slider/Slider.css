.arrow {
  display: flex;
  align-items: center;
}

.arrow button {
  width: 2.5rem;
  height: 2.5rem;
  margin: 0;
  border: 0;
  font-size: 2rem;
  font-weight: bold;
  background-color: white;
}


.slider-container {
  display: flex;
  margin: 0 auto;
  width: 100%;
  margin: 1rem 0;
}

.slider-content {
  display: flex;
  width: 100%;
  align-items: center;
}

.slider-item {
  width: 33.3%;
  border: 1px solid;
  font-size: 1.25rem;
  line-height: 8rem;
  margin: 10px;
  padding: 2%;
  position: relative;
  text-align: center;
}
