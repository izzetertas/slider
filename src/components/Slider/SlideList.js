import React from 'react'
import './Slider.css'

const SlideList = ({ items }) => {
  if(!items) return null

  return (
    <div className='slider-content'>
    {
      items.map((item, i) => (
        <div className='slider-item' key={i}>
          {item}
        </div>
      ))
    }
    </div>
  )
}

export default SlideList
