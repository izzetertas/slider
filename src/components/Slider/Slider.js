import React, { Component } from 'react'
import PropTypes from 'prop-types'
import SlideList from './SlideList'
import './Slider.css'

class Slider extends Component {
  state = {
    items : this.props.items || [],
    firstActiveSlide: 0
  }

  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.string)
  }

  static getDerivedStateFromProps(nextProps, prevState){
    if(nextProps.items && nextProps.items.length !== prevState.items.length) {
      return {
        items: nextProps.items
      }
    }
    return {}
  }

  get isNextButtonActive() {
    const { firstActiveSlide, items } = this.state
    return items.length < 4 || firstActiveSlide + 2 === items.length-1
  }

  get isPreviousButtonActive() {
    return this.state.firstActiveSlide === 0
  }

  getActiveSlides() {
    const { items } = this.state
    if(!items) return null

    const { firstActiveSlide } = this.state
    return items.filter((item, i) => i >= firstActiveSlide && i <= firstActiveSlide + 2)
  }

  renderArrow(isDisabled, count, text){
    return (
      <div className='arrow'>
        <button
          disabled={isDisabled}
          onClick={() => this.setState({ firstActiveSlide: this.state.firstActiveSlide + count})}
        >
          {text}
        </button>
      </div>
    )
  }

  render() {
    const items = this.getActiveSlides()

    return(
      <div className='slider-container'>
        {this.renderArrow(this.isPreviousButtonActive, -1, '<')}
        <SlideList items={items}/>
        {this.renderArrow(this.isNextButtonActive, 1, '>')}
      </div>
    )
  }
}

export default Slider