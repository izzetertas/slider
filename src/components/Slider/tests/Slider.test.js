import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import Slider from '../Slider'

let wrapper;

describe('<Slider/>', () => {
  beforeEach(() => {
    // initialState = {
    //   items: []
    // }
    wrapper = shallow(<Slider  />)
  })
  it('should render correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('button').length).toBe(2)
    expect(wrapper.find('SlideList').length).toBe(1)  
  });

  describe('When slide items are less than 3', () => {
    it('arrow buttons should be disabled', () => {
      wrapper = shallow(<Slider items={['1', '2']} />)
      expect(wrapper.find('button').find({disabled: true}).length).toBe(2)
    });
  });

  describe('When slide items count is equal 4 and firstActiveSlide index is equal 1', () => {
    beforeEach(() => {
      wrapper = shallow(<Slider items={['1', '2', '3', '4']} />)
      wrapper.setState({ firstActiveSlide: 1})
    })
    it('left button should be enabled', () => {
      expect(wrapper.find('button').first().prop('disabled')).toBeFalsy()
    });
    it('right button should be disabled', () => {
      expect(wrapper.find('button').at(1).prop('disabled')).toBeTruthy()
    });
  });

  describe('When slide items count is equal 5 and firstActiveSlide index is equal 0', () => {
    beforeEach(() => {
      wrapper = shallow(<Slider items={['1', '2', '3', '4', '5']} />)
      wrapper.setState({ firstActiveSlide: 0})
    })
    it('left button should be disabled', () => {
      expect(wrapper.find('button').first().prop('disabled')).toBeTruthy()
    });
    it('right button should be disabled', () => {
      expect(wrapper.find('button').at(1).prop('disabled')).toBeFalsy()
    });
  });

  describe('When new slides comes from props', () => {
    it('should update state', () => {
      wrapper = shallow(<Slider items={['1']}/>)
      const newItems = ['1', '2', '3']
      wrapper.setProps({ items : newItems})
      expect(wrapper.state('items')).toEqual(newItems)
    });
  });

  describe('When arrow botton is clicked', () => {
    it('should update firstActiveSlide index', () => {
      wrapper = shallow(<Slider items={['1', '2', '3', '4', '5']} />)
      wrapper.setState({ firstActiveSlide: 0})
      wrapper.find('button').at(1).simulate('click')
      expect(wrapper.state('firstActiveSlide')).toEqual(1)

      wrapper.find('button').at(0).simulate('click')
      expect(wrapper.state('firstActiveSlide')).toEqual(0)
    });
  });
})