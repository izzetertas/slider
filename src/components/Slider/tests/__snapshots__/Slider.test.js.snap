// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider/> should render correctly 1`] = `
<div
  className="slider-container"
>
  <div
    className="arrow"
  >
    <button
      disabled={true}
      onClick={[Function]}
    >
      &lt;
    </button>
  </div>
  <SlideList
    items={Array []}
  />
  <div
    className="arrow"
  >
    <button
      disabled={true}
      onClick={[Function]}
    >
      &gt;
    </button>
  </div>
</div>
`;
